From: Russell King <rmk+kernel@armlinux.org.uk>
Date: Thu, 5 Jan 2017 12:21:09 +0000
Subject: [PATCH] net: phy: restart phy autonegotiation after EEE
 advertisment change

When the EEE advertisment is changed, we should restart autonegotiation
to update the link partner with the new EEE settings.  Add this trigger
but only if the advertisment has changed.

Signed-off-by: Russell King <rmk+kernel@armlinux.org.uk>
---

--- a/drivers/net/phy/phy.c
+++ b/drivers/net/phy/phy.c
@@ -1343,6 +1343,7 @@ EXPORT_SYMBOL(phy_ethtool_get_eee);
  */
 int phy_ethtool_set_eee(struct phy_device *phydev, struct ethtool_eee *data)
 {
+	int ret;
 	int val = ethtool_adv_to_mmd_eee_adv_t(data->advertised);
 
 	/* Mask prohibited EEE modes */
@@ -1350,6 +1351,15 @@ int phy_ethtool_set_eee(struct phy_devic
 
 	phy_write_mmd_indirect(phydev, MDIO_AN_EEE_ADV, MDIO_MMD_AN, val);
 
+	if (phydev->eee_broken_modes) {
+		/* Restart autonegotiation so the new modes get sent to the
+		 * link partner.
+		 */
+		ret = phy_restart_aneg(phydev);
+		if (ret < 0)
+			return ret;
+	}
+
 	return 0;
 }
 EXPORT_SYMBOL(phy_ethtool_set_eee);
